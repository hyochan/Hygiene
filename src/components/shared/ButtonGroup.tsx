import React, { useState } from 'react';
import {
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';

interface Props {
  testID?: string;
  borderRadius?: number;
  containerStyle?: StyleProp<ViewStyle>;
  style?: StyleProp<ViewStyle>;
  viewStyle?: StyleProp<ViewStyle>;
  selectedViewStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  selectedTextStyle?: StyleProp<TextStyle>;
  data: string[];
  onPress?: (i: number) => void;
  initialIndex?: number;
}

function Shared(props: Props): React.ReactElement {
  const {
    borderRadius = 0,
    initialIndex = 0,
    testID,
    containerStyle,
    style,
    selectedViewStyle,
    viewStyle,
    selectedTextStyle,
    textStyle,
    data,
    onPress,
  } = props;

  const [selectedOption, setSelectedOption] = useState(initialIndex);

  return (
    <View testID={testID} style={StyleSheet.flatten([containerStyle, style])}>
      {data.map((text, i) => {
        return (
          <TouchableOpacity
            key={i}
            testID={`CHILD_${i}`}
            style={{ flex: 1 }}
            onPress={(): void => {
              setSelectedOption(i);

              if (onPress) {
                onPress(i);
              }
            }}
          >
            <View
              style={
                StyleSheet.flatten([
                  selectedOption === i
                    ? selectedViewStyle
                    : viewStyle,
                  i === 0 ? {
                    borderTopLeftRadius: borderRadius,
                    borderBottomLeftRadius: borderRadius,
                  } : {},
                  i === data.length - 1 ? {
                    borderBottomRightRadius: borderRadius,
                    borderTopRightRadius: borderRadius,
                  } : {},
                ])
              }
            >
              <Text
                style={
                  selectedOption === i
                    ? selectedTextStyle
                    : textStyle
                }
              >
                {text}
              </Text>
            </View>
          </TouchableOpacity>
        );
      })}
    </View>
  );
}

Shared.defaultProps = {
  containerStyle: {
    backgroundColor: 'transparent',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    borderWidth: 1,
    borderColor: 'rgb(62,126,255)',
    alignSelf: 'stretch',
    minHeight: 40,
    marginTop: 24,
  },
  viewStyle: {
    backgroundColor: 'white',
    alignSelf: 'stretch',
    minHeight: 40,
    borderColor: 'rgb(62,126,255)',

    justifyContent: 'center',
    alignItems: 'center',
  },
  selectedViewStyle: {
    backgroundColor: 'rgb(62,126,255)',
    alignSelf: 'stretch',
    minHeight: 40,
    borderColor: 'rgb(62,126,255)',

    justifyContent: 'center',
    alignItems: 'center',
  },
  textStyle: {
    color: 'rgb(62,126,255)',
    fontSize: 14,
    textAlign: 'center',
    alignSelf: 'center',
  },
  selectedTextStyle: {
    color: 'white',
    fontSize: 14,
    textAlign: 'center',
    alignSelf: 'center',
  },
  selectedOption: {
    index: 0,
  },
  data: ['option 1', 'option 2'],
};

export default Shared;
