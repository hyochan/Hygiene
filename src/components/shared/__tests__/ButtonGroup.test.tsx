import 'react-native';

import * as React from 'react';

import { RenderAPI, act, fireEvent, render } from '@testing-library/react-native';

import ButtonGroup from '../ButtonGroup';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let props: any;
let component: React.ReactElement;
let testingLib: RenderAPI;

const createTestProps = (obj: Record<string, unknown>): Record<string, unknown> => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...obj,
});

describe('[ButtonGroup] render', () => {
  beforeEach(() => {
    props = createTestProps({});
    component = <ButtonGroup {...props} />;
  });

  it('renders without crashing', () => {
    const rendered: renderer.ReactTestRendererJSON = renderer
      .create(component)
      .toJSON();

    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  describe('interactions', () => {
    beforeEach(() => {
      testingLib = render(component);
    });

    it('should simulate onPress', () => {
      const btn1 = testingLib.queryByTestId('CHILD_1');

      act(() => {
        fireEvent.press(btn1);
      });
    });
  });
});
