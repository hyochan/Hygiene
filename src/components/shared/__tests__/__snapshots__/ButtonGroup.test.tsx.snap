// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ButtonGroup] render renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "backgroundColor": "transparent",
      "borderColor": "rgb(62,126,255)",
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginTop": 24,
      "minHeight": 40,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "opacity": 1,
      }
    }
    testID="CHILD_0"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "stretch",
          "backgroundColor": "rgb(62,126,255)",
          "borderBottomLeftRadius": 0,
          "borderColor": "rgb(62,126,255)",
          "borderTopLeftRadius": 0,
          "justifyContent": "center",
          "minHeight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "white",
            "fontSize": 14,
            "textAlign": "center",
          }
        }
      >
        option 1
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "opacity": 1,
      }
    }
    testID="CHILD_1"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "stretch",
          "backgroundColor": "white",
          "borderBottomRightRadius": 0,
          "borderColor": "rgb(62,126,255)",
          "borderTopRightRadius": 0,
          "justifyContent": "center",
          "minHeight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "rgb(62,126,255)",
            "fontSize": 14,
            "textAlign": "center",
          }
        }
      >
        option 2
      </Text>
    </View>
  </View>
</View>
`;
